src/Microsoft.AspNetCore.SignalR/TypedClientBuilder.cs

```diff
@@ -0,0 +1,243 @@
+        public static T Build(IClientProxy proxy)
+        {
+            return _builder.Value(proxy);
+        }
+
+        [SuppressMessage("Microsoft.Performance", "CA1804:RemoveUnusedLocals", MessageId = "forceEvaluation", Justification = "Give me a better way to force the evaluation of Lazy<T>")]
+        public static void Validate()
+        {
+            // The following will throw if T is not a valid type
+            var forceEvaluation = _builder.Value;
+        }
+
+        private static Func<IClientProxy, T> GenerateClientBuilder()
+        {
+            VerifyInterface(typeof(T));
+
+            var assemblyName = new AssemblyName(ClientModuleName);
+            AssemblyBuilder assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(assemblyName, AssemblyBuilderAccess.Run);
+            ModuleBuilder moduleBuilder = assemblyBuilder.DefineDynamicModule(ClientModuleName);
+            Type clientType = GenerateInterfaceImplementation(moduleBuilder);
+
+            return proxy => (T)Activator.CreateInstance(clientType, proxy);
+        }
+
+        private static Type GenerateInterfaceImplementation(ModuleBuilder moduleBuilder)
+        {
+            TypeBuilder type = moduleBuilder.DefineType(
+                ClientModuleName + "." + typeof(T).Name + "Impl",
+                TypeAttributes.Public,
+                typeof(Object),
+                new[] { typeof(T) });
+
+            FieldBuilder proxyField = type.DefineField("_proxy", typeof(IClientProxy), FieldAttributes.Private);
+
+            BuildConstructor(type, proxyField);
+
+            foreach (var method in GetAllInterfaceMethods(typeof(T)))
+            {
+                BuildMethod(type, method, proxyField);
+            }
+
+            return type.CreateTypeInfo();
+        }
+
+        private static IEnumerable<MethodInfo> GetAllInterfaceMethods(Type interfaceType)
+        {
+            foreach (var parent in interfaceType.GetInterfaces())
+            {
+                foreach (var parentMethod in GetAllInterfaceMethods(parent))
+                {
+                    yield return parentMethod;
+                }
+            }
+
+            foreach (var method in interfaceType.GetMethods())
+            {
+                yield return method;
+            }
+        }
+
+        private static void BuildConstructor(TypeBuilder type, FieldInfo proxyField)
+        {
+            MethodBuilder method = type.DefineMethod(".ctor", System.Reflection.MethodAttributes.Public | System.Reflection.MethodAttributes.HideBySig);
+
+            ConstructorInfo ctor = typeof(object).GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic,
+                null, new Type[] { }, null);
+
+            method.SetReturnType(typeof(void));
+            method.SetParameters(typeof(IClientProxy));
+
+            ILGenerator generator = method.GetILGenerator();
+
+            // Call object constructor
+            generator.Emit(OpCodes.Ldarg_0);
```

Comment:
When writing IL generation code, it's important to track what the stack looks like. I'd suggest adding comments to these lines indicating the stack contents and any opcode effects after each instruction. IL is tricky to follow, and it's much easier if you can just verify that each line does what the comment says, and then verify that the flow of data in the comments makes sense.

```csharp
generator.Emit(OpCodes.Ldarg_0); // Stack: ... arg0
generator.Emit(OpCodes.Call, ctor); // base(); Stack: ...
generator.Emit(OpCodes.Ldarg_0); // Stack: ... arg0
generator.Emit(OpCodes.Ldarg_1); // Stack: ... arg1 arg0
generator.Emit(OpCodes.Stfld); // arg0.proxyField = arg1; Stack: ...

// etc...
```